jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c

```diff
@@ -574,266 +472,301 @@ typedef enum
+  /* Handle infinity separately */
+  if (ecma_number_is_infinity (this_num))
   {
-    return comp_value;
+    ecma_stringbuilder_append_magic (&builder, LIT_MAGIC_STRING_INFINITY_UL);
+    return ecma_make_string_value (ecma_stringbuilder_finalize (&builder));
   }
 
-  /* Get the parameters of the number if non-zero. */
-  lit_utf8_byte_t digits[ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER];
-  lit_utf8_size_t num_digits;
-  int32_t exponent;
+  num_of_digits = ecma_number_to_decimal (this_num, digits, &exponent);
 
-  if (!ecma_number_is_zero (this_num))
-  {
-    num_digits = ecma_number_to_binary_floating_point_number (this_num, digits, &exponent);
-    JERRY_ASSERT (exponent >= 0);
-  }
-  else
+  /* Handle undefined argument */
+  if (ecma_is_value_undefined (arg))
   {
-    for (int32_t i = 0; i <= frac_digits; i++)
+    if (mode == NUMBER_ROUTINE_TO_FIXED)
     {
-      digits[i] = '0';
+      arg_int = 0;
+    }
+    else if (mode == NUMBER_ROUTINE_TO_EXPONENTIAL)
+    {
+      arg_int = (int32_t) num_of_digits - 1;
     }
-    num_digits = (lit_utf8_size_t) frac_digits + 1;
-    exponent = 1;
   }
 
-  /* 7. */
-  if (exponent > 21)
+  if (mode == NUMBER_ROUTINE_TO_FIXED
+      && exponent > 21)
   {
+    ecma_stringbuilder_destroy (&builder);
     return ecma_builtin_number_prototype_object_to_string (this_num, NULL, 0);
   }
 
-  /* 8. */
-  num_digits = ecma_builtin_number_prototype_helper_round (digits,
-                                                           num_digits + (lit_utf8_size_t) exponent,
-                                                           exponent + frac_digits,
-                                                           &exponent,
-                                                           ecma_number_is_zero (this_num));
+  int32_t digits_to_keep = arg_int;
 
-  /* Buffer that is used to construct the string. */
-  int buffer_size = (exponent > 0) ? exponent + frac_digits + 2 : frac_digits + 3;
-
-  if (is_negative)
+  if (mode == NUMBER_ROUTINE_TO_FIXED)
   {
-    buffer_size++;
+    digits_to_keep += exponent;
+  }
+  else if (mode == NUMBER_ROUTINE_TO_EXPONENTIAL)
+  {
+    digits_to_keep += 1;
   }
 
-  JERRY_ASSERT (buffer_size > 0);
-
-  ecma_value_t ret_value;
-  JMEM_DEFINE_LOCAL_ARRAY (buff, buffer_size, lit_utf8_byte_t);
-
-  lit_utf8_byte_t *p = buff;
+  num_of_digits = ecma_builtin_number_prototype_helper_round (digits,
+                                                              num_of_digits,
+                                                              digits_to_keep,
+                                                              &exponent,
+                                                              false);
 
-  if (is_negative)
+  /* toExponent routine and toPrecision cases where the exponent > precision or exponent < -5 */
+  if (mode == NUMBER_ROUTINE_TO_EXPONENTIAL
+      || (mode == NUMBER_ROUTINE_TO_PRECISION
+          && (exponent < -5 || exponent > arg_int)))
```

Comment:
This range seems to be different in the standard. Please add tests that cover the edge cases and make sure they are correct.